arch: riscv: Fix warning when C++ is enabled

thughes opened the pull request:
When compiling with C++ enabled (CONFIG_CPP), add an unused member to
prevent an empty struct; this makes the struct size the same for both C
and C++.

Fixes the following warnings:

In file included from include/zephyr/drivers/gpio.h:22:
In file included from include/zephyr/tracing/tracing.h:9:
In file included from include/zephyr/kernel.h:17:
In file included from include/zephyr/kernel_includes.h:32:
In file included from include/zephyr/kernel_structs.h:29:
In file included from include/zephyr/arch/structs.h:29:
include/zephyr/arch/riscv/structs.h:11:1: error: empty struct has size 0
in C, size 1 in C++ [-Werror,-Wextern-c-compat]
   11 | struct _cpu_arch {
      | ^

In file included from include/zephyr/drivers/gpio.h:22:
In file included from include/zephyr/tracing/tracing.h:9:
In file included from include/zephyr/kernel.h:17:
In file included from include/zephyr/kernel_includes.h:36:
In file included from include/zephyr/arch/cpu.h:25:
In file included from include/zephyr/arch/riscv/arch.h:18:
include/zephyr/arch/riscv/thread.h:68:1: error: empty struct has size 0
in C, size 1 in C++ [-Werror,-Wextern-c-compat]
   68 | struct _thread_arch {
      | ^

thughes committed:
arch: riscv: Fix warning when C++ is enabled

When compiling with C++ enabled (CONFIG_CPP), add an unused member to
prevent an empty struct; this makes the struct size the same for both C
and C++.

Fixes the following warnings:

In file included from include/zephyr/drivers/gpio.h:22:
In file included from include/zephyr/tracing/tracing.h:9:
In file included from include/zephyr/kernel.h:17:
In file included from include/zephyr/kernel_includes.h:32:
In file included from include/zephyr/kernel_structs.h:29:
In file included from include/zephyr/arch/structs.h:29:
include/zephyr/arch/riscv/structs.h:11:1: error: empty struct has size 0
in C, size 1 in C++ [-Werror,-Wextern-c-compat]
   11 | struct _cpu_arch {
      | ^

In file included from include/zephyr/drivers/gpio.h:22:
In file included from include/zephyr/tracing/tracing.h:9:
In file included from include/zephyr/kernel.h:17:
In file included from include/zephyr/kernel_includes.h:36:
In file included from include/zephyr/arch/cpu.h:25:
In file included from include/zephyr/arch/riscv/arch.h:18:
include/zephyr/arch/riscv/thread.h:68:1: error: empty struct has size 0
in C, size 1 in C++ [-Werror,-Wextern-c-compat]
   68 | struct _thread_arch {
      | ^

Signed-off-by: Tom Hughes <tomhughes@chromium.org>

fkokosinski commented:
Hi, thanks for this PR! Unfortunately, I can't seem to reproduce this warning using the following command:
```
% ~ west build -p -b qemu_riscv32 samples/philosophers -- -DCONFIG_CPP=y
```

Could you perhaps create an issue with information on how to reproduce this warning?

thughes commented:
> Could you perhaps create an issue with information on how to reproduce this warning?

There are two things in my setup that are different than the way you're testing:

* I'm using LLVM (clang version 20).
* I'm building a board that hasn't yet been integrated into upstream Zephyr: https://crrev.com/c/6349236.

It should be reproducible if you use LLVM and make sure that the relevant configs are disabled that make `struct _cpu_arch` and ` struct _thread_arch` empty.

Here's an example of what the failure looks like in CI when building with LLVM in another part of the code: https://github.com/zephyrproject-rtos/zephyr/pull/87088#discussion_r2049295801

fkokosinski commented:
> There are two things in my setup that are different than the way you're testing:
> * I'm using LLVM (clang version 20).

Right. This is a configuration that, as far as I’m aware, is unsupported by Zephyr out of the box.

> * I'm building a board that hasn't yet been integrated into upstream Zephyr:[ https://crrev.com/c/6349236](https://crrev.com/c/6349236).

Unfortunately, I can’t access the link without signing in, but I assume you had to at least add a CMake script in `cmake/compiler/clang/riscv.cmake`. Would it be possible for you to extract that part, and contribute it as well as a part of this PR? And preferably test it with a platform that has at least one entry in `SUPPORTED_EMU_PLATFORMS`?

This is something we’re not testing right now in CI, and I’d rather we add this configuration to our suite as well and make it much more easily accessible for maintainers to test and use.

thughes commented:
> This is a configuration that, as far as I’m aware, is unsupported by Zephyr out of the box.

LLVM is officially supported and tested with CI: https://github.com/zephyrproject-rtos/zephyr/issues/85542. For example, see https://github.com/zephyrproject-rtos/zephyr/pull/87088#discussion_r2049295801. However, the CI is currently only building for `native-sim`.

> Unfortunately, I can’t access the link without signing in

https://crrev.com/c/6349236 is publicly accessible. No sign in is required to view it.

I'll see if I can create an upstream change to cover this path in CI. At the very least, it should be possible to demonstrate the size issue with gcc using `BUILD_ASSERT(sizeof(struct _thread_arch) >= 1);`.

thughes committed:
tests: riscv: Add test to verify C++ compilation

Using clang to build for RISC-V with these configs:

CONFIG_CPP=y
CONFIG_USERSPACE=n
CONFIG_SMP=n
CONFIG_FPU_SHARING=n

results in the following warnings:

include/zephyr/arch/riscv/structs.h:11:1: error: empty struct has size 0
in C, size 1 in C++ [-Werror,-Wextern-c-compat]
11 | struct _cpu_arch {
| ^

include/zephyr/arch/riscv/thread.h:68:1: error: empty struct has size 0
in C, size 1 in C++ [-Werror,-Wextern-c-compat]
68 | struct _thread_arch {
| ^

This commit adds a test with those configs to verify compilation
succeeds. Although gcc doesn't emit these warnings, we can test with gcc
as well with a BUILD_ASSERT in include/zephyr/arch/riscv/thread.h and
include/zephyr/arch/riscv/structs.h.

Signed-off-by: Tom Hughes <tomhughes@chromium.org>

thughes commented:
@fkokosinski I added another commit that adds a `build_only` test case. You can see it fail with `gcc` by commenting out the struct change in `include/zephyr/arch/riscv/structs.h`, while leaving the `BUILD_ASSERT(sizeof(struct _cpu_arch) >= 1);` and running it:

```
./scripts/twister -c -p qemu_riscv32 -s arch.riscv.test_build_cpp
```

For reference, here's another change in the code where we did the same thing: https://github.com/zephyrproject-rtos/zephyr/pull/88778.

ycsin commented on include/zephyr/arch/riscv/thread.h:
```diff
@@ -84,8 +84,17 @@ struct _thread_arch {
 	unsigned long m_mode_pmpaddr_regs[PMP_M_MODE_SLOTS];
 	unsigned long m_mode_pmpcfg_regs[PMP_M_MODE_SLOTS / sizeof(unsigned long)];
 #endif
+#if defined(CONFIG_CPP) && !defined(CONFIG_FPU_SHARING) && !defined(CONFIG_USERSPACE) &&           \
+	!defined(CONFIG_PMP_STACK_GUARD)
+	/* Empty struct has size 0 in C, size 1 in C++. Force them to be the same. */
+	uint8_t unused_cpp_size_compatibility;
+#endif
```
I wonder if we should have something like the following defined somewhere

```c
#ifdef __cplusplus
#define CPP_DUMMY_STRUCT_MEMBER int __dummy__
#else
#define CPP_DUMMY_STRUCT_MEMBER
#endif
```

and used in structs that can be empty

```c
struct _thread_arch {
#ifdef CONFIG_FPU_SHARING
	struct z_riscv_fp_context saved_fp_context;
	bool fpu_recently_used;
	uint8_t exception_depth;
#endif
#ifdef CONFIG_USERSPACE
	unsigned long priv_stack_start;
	unsigned long u_mode_pmpaddr_regs[CONFIG_PMP_SLOTS];
	unsigned long u_mode_pmpcfg_regs[CONFIG_PMP_SLOTS / sizeof(unsigned long)];
	unsigned int u_mode_pmp_domain_offset;
	unsigned int u_mode_pmp_end_index;
	unsigned int u_mode_pmp_update_nr;
#endif
#ifdef CONFIG_PMP_STACK_GUARD
	unsigned int m_mode_pmp_end_index;
	unsigned long m_mode_pmpaddr_regs[PMP_M_MODE_SLOTS];
	unsigned long m_mode_pmpcfg_regs[PMP_M_MODE_SLOTS / sizeof(unsigned long)];
#endif
	CPP_DUMMY_STRUCT_MEMBER;
}
```


so that the code looks cleaner, at the expense of an extra `int` for every such struct when compiled for cpp, regardless of whether or not they are actually empty.

fkokosinski replied:
I like the idea of the dummy struct member, but I think you got the condition backwards? Plus, the dummy member should probably be a `char` instead.

We could avoid the dummy member this way:
```C
#ifdef __cplusplus
#define C_DUMMY_STRUCT_MEMBER
#else
#define C_DUMMY_STRUCT_MEMBER char __dummy__
#endif

struct foobar {
#ifdef CONFIG_FOO
	int foo;
#endif
#ifdef CONFIG_BAR
	int bar;
#endif
#if !(CONFIG_FOO || CONFIG_BAR)
	C_DUMMY_STRUCT_MEMBER;
#endif
};
```

thughes replied:
How about I implement something like this in a followup PR?

ycsin replied:
> but I think you got the condition backwards?

No really, I think we are talking about different idea - I'm proposing an alternative where we _always_ have a dummy variable in every struct that might be empty when `CONFIG_CPP=y`
```c
#ifdef __cplusplus
#define CPP_DUMMY_STRUCT_MEMBER char __dummy__ <- make sure that a struct can never be empty when CONFIG_CPP=y
#else
#define CPP_DUMMY_STRUCT_MEMBER
#endif
```
and whenever we have a struct that can be empty due to Kconfigs, simply append `CPP_DUMMY_STRUCT_MEMBER` at the end to avoid more preprocessor macros like these:
```c
#if !(CONFIG_FOO || CONFIG_BAR)
        CPP_DUMMY_STRUCT_MEMBER;
#endif
```

fkokosinski commented:
> @fkokosinski I added another commit that adds a `build_only` test case. You can see it fail with `gcc` by commenting out the struct change in `include/zephyr/arch/riscv/structs.h`, while leaving the `BUILD_ASSERT(sizeof(struct _cpu_arch) >= 1);` and running it:
> 
> ```
> ./scripts/twister -c -p qemu_riscv32 -s arch.riscv.test_build_cpp
> ```
> 
> For reference, here's another change in the code where we did the same thing: [#87088 (review)](https://github.com/zephyrproject-rtos/zephyr/pull/87088#pullrequestreview-2776330686) and #88778.

Thanks!

ycsin commented on include/zephyr/arch/riscv/structs.h:
```diff
@@ -22,6 +22,15 @@ struct _cpu_arch {
+#endif
 };
 
+#if defined(CONFIG_CPP)
+BUILD_ASSERT(sizeof(struct _cpu_arch) >= 1);
+#endif
```
What's the purpose of this? To produce a different compilation error?

thughes replied:
It makes the build failure compiler agnostic. `gcc` doesn't generate this warning, but `clang` does.

ycsin commented on tests/arch/riscv/test_build_cpp/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+
+cmake_minimum_required(VERSION 3.20.0)
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(test_build_cpp)
```
If we are going to add such test, it should probably be arch-agnostic

thughes replied:
See latest changes.

ycsin reviewed (commented):
Probably a RFC to streamline how we would deal with possibly empty struct in the future, but +1 to this PR

thughes commented on tests/lib/cpp/cxx/testcase.yaml:
```diff
@@ -73,3 +73,11 @@ tests:
       - CONFIG_STD_CPP2B=y
+  # Verify struct _cpu_arch struct _thread_arch are compatible with C++ when
+  # specific configs are enabled.
+  cpp.main.empty_struct_size:
```
The test is arch agnostic, but the configs required to trigger the empty struct are architecture dependent. We could put an `arch_allow: riscv` on the test to make that clear. What do you think?

thughes commented:
I can't seem to re-open this, so I created https://github.com/zephyrproject-rtos/zephyr/pull/96943.

thughes commented:
Test moved from this PR to https://github.com/zephyrproject-rtos/zephyr/pull/96944
